File: sorting_algorithm/insertion_sort.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
"""""""""""""""""""""""""""""""""""""""""""""""
"      Filename: insertion_sort.py
"
"        Author: xss - callmexss@126.com
"   Description: ---
"        Create: 2017-12-06 22:07:49
"""""""""""""""""""""""""""""""""""""""""""""""
import pyhandy.pyhandy as handy


def insertion_sort(num):
    # num: list of number
    if len(num) == 1:
        return num

    for i in range(1, len(num)):
        temp = num[i]
        j = i

        # find a suitable location for num[i]
        # so move each element in sorted part to its next location
        # if it is larger than num[i]
        # and put num[i] after the first element smaller than it
        while(j > 0 and (num[j] < num[j-1])):
            num[j], num[j-1] = num[j-1], num[j]
            j -= 1

        num[j] = temp

    return num


if __name__ == "__main__":
    rh = handy.RandomHandy()
    sh = handy.SortHandy()
    n = 10000
    sh.testSort(insertion_sort, rh.generate_nearly_ordered_array(n))
    sh.testSort(insertion_sort, rh.generate_random_array(n, 0, n))

File: sorting_algorithm/insertion_sort.cpp
/***********************************************
 *
 *      Filename: insertion_sort.cpp
 *
 *        Author: xss - callmexss@126.com
 *   Description: insertion sort
 *        Create: 2017-12-06 21:28:36
 *
 ***********************************************/

#include <iostream>

#include "sorttesthelper.h"

using namespace std;

template <typename T>
void InsertionSort(T arr[], int len)
{
    int i, j;
    T temp;

    for (i = 1; i < len; i++)
    {
        temp = arr[i];

        for (j = i; j > 0 && arr[j - 1] > temp; j--)
        {
            arr[j] = arr[j - 1];
        }

        arr[j] = temp;
    }
}

template <typename T>
void Insert(T arr[], int len)
{
    int i;
    T temp;
    temp = arr[len];

    for (i = len - 1; i >= 0 && temp < arr[i]; i--)
    {
        arr[i + 1] = arr[i];
    }
    arr[i + 1] = temp;
}

template <typename T>
void InsertionSortRecursion(T arr[], int len)
{
    if (len > 0)
    {
        InsertionSortRecursion(arr, len - 1);
        Insert(arr, len);
    }
}

int main(int argc, const char *argv[])
{
    // int num[] = {3, 1, 4, 9, 2, 6, 5, 3};
    //InsertionSort(num, 8);
    int n = 100000;
    int *num = SortTestHelper::generateNearlyOrderedArray(n);
    // InsertionSortRecursion(num, n);

    // SortTestHelper::printArray(num, n);

    SortTestHelper::testSort("Insertion Sort", InsertionSort, num, n);

    delete[] num;

    return 0;
}

// reference: http://www.cnblogs.com/zhuorongtan/archive/2012/09/22/2698339.html
